components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -182,12 +196,71 @@ abstract class AbstractFetchDownloadService : Service() {
             startDownloadJob(download.id)
         }
 
+        notificationUpdateScope.launch {
+            while (isActive) {
+                delay(PROGRESS_UPDATE_INTERVAL)
+                updateDownloadNotification()
+            }
+        }
+
         return super.onStartCommand(intent, flags, startId)
     }
 
+    /***
+     * Android rate limits notifications being sent, so we must send them on a delay so that
+     * notifications are not dropped
+     */
+    private fun updateDownloadNotification() {
+        synchronized(context) {
+            for (download in downloadJobs.values) {
+                // Dispatch the corresponding notification based on the current status
+                val notification = when (download.status) {
+                    DownloadJobStatus.ACTIVE -> {
+                        synchronized(context) {
```

Comment:
It looks like you are already in a synchronized block here (the whole method is?).